File: pdf/golang1/controllers/pdf_controller.go
package controllers

import (
	"fmt"
	"net/http"
	"note-pdf-golang/services"

	"github.com/labstack/echo/v4"
)

type PdfController interface {
	GetPdf(c echo.Context) error
}

type pdfController struct {
	PdfService services.PdfService
}

func NewPdfController(pdfService services.PdfService) PdfController {
	return &pdfController{
		PdfService: pdfService,
	}
}

func (controller *pdfController) GetPdf(c echo.Context) error {
	pdfBytes, err := controller.PdfService.GetPdf()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]interface{}{
			"response": err.Error(),
		})
	}
	c.Response().Header().Set("Content-Type", "application/pdf")
	c.Response().Header().Set("Content-Disposition", "attachment; filename=output.pdf")
	fmt.Println("pdfBytes.Bytes():", pdfBytes.Bytes())
	c.Response().Write(pdfBytes.Bytes())
	return nil
}

File: pdf/rust/src/controllers/pdf_controller.rs
use std::sync::Arc;

use axum::{extract::State, http::{HeaderMap, HeaderValue, StatusCode}, response::{IntoResponse, Response}};

use crate::{services::pdf_service::PdfService, states::app_state::AppState};

pub async fn generate_pdf(
    State(state): State<Arc<AppState>>
) -> Response {
    match state.pdf_service.generate_pdf_from_string().await {
        Ok(pdf_bytes) => {
            let mut headers = HeaderMap::new();
            headers.insert("Content-Type", HeaderValue::from_static("application/pdf"));
            headers.insert("Content-Disposition", HeaderValue::from_static("attachment; filename=\"output.pdf\""));
            (headers, pdf_bytes).into_response()
        }
        Err(err) => {
            eprintln!("PDF generation error: {:?}", err);
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                "Failed to generate pdf"
            ).into_response()
        }
    }
}